File: android/app/src/main/java/com/purchasely/shaker/ui/screen/home/HomeViewModel.kt
package com.purchasely.shaker.ui.screen.home

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.purchasely.shaker.domain.repository.CocktailRepository
import com.purchasely.shaker.domain.repository.PremiumRepository
import com.purchasely.shaker.domain.model.Cocktail
import com.purchasely.shaker.domain.model.CocktailMood
import com.purchasely.shaker.domain.usecase.GetFilteredCocktailsUseCase
import com.purchasely.shaker.purchasely.FetchResult
import com.purchasely.shaker.purchasely.PresentationHandle
import com.purchasely.shaker.purchasely.PurchaselyWrapper
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class HomeViewModel(
    private val repository: CocktailRepository,
    private val premiumRepository: PremiumRepository,
    private val purchaselyWrapper: PurchaselyWrapper,
    private val getFilteredCocktails: GetFilteredCocktailsUseCase
) : ViewModel() {

    private val _cocktails = MutableStateFlow<List<Cocktail>>(emptyList())
    val cocktails: StateFlow<List<Cocktail>> = _cocktails.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    val isPremium: StateFlow<Boolean> = premiumRepository.isPremium

    // Filter state
    private val _selectedSpirits = MutableStateFlow<Set<String>>(emptySet())
    val selectedSpirits: StateFlow<Set<String>> = _selectedSpirits.asStateFlow()

    private val _selectedCategories = MutableStateFlow<Set<String>>(emptySet())
    val selectedCategories: StateFlow<Set<String>> = _selectedCategories.asStateFlow()

    private val _selectedDifficulty = MutableStateFlow<String?>(null)
    val selectedDifficulty: StateFlow<String?> = _selectedDifficulty.asStateFlow()

    private val _selectedMood = MutableStateFlow<CocktailMood?>(null)
    val selectedMood: StateFlow<CocktailMood?> = _selectedMood.asStateFlow()

    val availableSpirits: List<String> get() = repository.getSpirits()
    val availableCategories: List<String> get() = repository.getCategories()
    val availableDifficulties: List<String> get() = repository.getDifficulties()

    private val _hasActiveFilters = MutableStateFlow(false)
    val hasActiveFilters: StateFlow<Boolean> = _hasActiveFilters.asStateFlow()

    private fun updateHasActiveFilters() {
        _hasActiveFilters.value = _selectedSpirits.value.isNotEmpty() ||
                _selectedCategories.value.isNotEmpty() ||
                _selectedDifficulty.value != null ||
                _selectedMood.value != null
    }

    // Prefetched inline presentation
    private val _inlinePresentation = MutableStateFlow<FetchResult?>(null)
    val inlinePresentation: StateFlow<FetchResult?> = _inlinePresentation.asStateFlow()

    // Prefetched filters presentation
    private val _filtersPresentation = MutableStateFlow<FetchResult?>(null)
    val filtersPresentation: StateFlow<FetchResult?> = _filtersPresentation.asStateFlow()

    private val _isFiltersLoading = MutableStateFlow(false)
    val isFiltersLoading: StateFlow<Boolean> = _isFiltersLoading.asStateFlow()

    // Signal Screen to display filters presentation
    private val _requestPresentationDisplay = MutableSharedFlow<PresentationHandle>()
    val requestPresentationDisplay: SharedFlow<PresentationHandle> = _requestPresentationDisplay.asSharedFlow()

    init {
        _cocktails.value = getFilteredCocktails()
        prefetchPresentations()
    }

    private fun prefetchPresentations() {
        if (isPremium.value) return
        viewModelScope.launch {
            _isFiltersLoading.value = true
            _filtersPresentation.value = purchaselyWrapper.loadPresentation("filters")
            _isFiltersLoading.value = false
        }
        viewModelScope.launch {
            _inlinePresentation.value = purchaselyWrapper.loadPresentation("inline")
        }
    }

    fun onSearchQueryChanged(query: String) {
        _searchQuery.value = query
        if (query.isNotBlank()) {
            purchaselyWrapper.setUserAttribute("has_used_search", true)
        }
        applyFilters()
    }

    fun onFilterClick() {
        if (isPremium.value) return
        when (val result = _filtersPresentation.value) {
            is FetchResult.Success -> {
                viewModelScope.launch { _requestPresentationDisplay.emit(result.handle) }
            }
            is FetchResult.Error -> {
                // PURCHASELY: the prefetch failed (e.g. offline at launch). Retry it so the
                // paywall becomes available instead of leaving the button dead forever.
                Log.w(TAG, "[Shaker] Filters paywall unavailable (${result.message}), retrying prefetch")
                retryFiltersPrefetch()
            }
            is FetchResult.Deactivated, is FetchResult.Client -> {
                // Placement disabled in the console or client-rendered — nothing to display.
                Log.d(TAG, "[Shaker] Filters paywall not displayable: $result")
            }
            null -> {
                // Prefetch still in flight — the screen shows the loader; ignore the tap.
                Log.d(TAG, "[Shaker] Filters paywall still loading")
            }
        }
    }

    private fun retryFiltersPrefetch() {
        if (_isFiltersLoading.value) return
        viewModelScope.launch {
            _isFiltersLoading.value = true
            _filtersPresentation.value = purchaselyWrapper.loadPresentation("filters")
            _isFiltersLoading.value = false
        }
    }

    fun toggleSpirit(spirit: String) {
        val current = _selectedSpirits.value.toMutableSet()
        if (current.contains(spirit)) current.remove(spirit) else current.add(spirit)
        _selectedSpirits.value = current
        applyFilters()
        updateHasActiveFilters()
    }

    fun toggleCategory(category: String) {
        val current = _selectedCategories.value.toMutableSet()
        if (current.contains(category)) current.remove(category) else current.add(category)
        _selectedCategories.value = current
        applyFilters()
        updateHasActiveFilters()
    }

    fun selectDifficulty(difficulty: String?) {
        _selectedDifficulty.value = if (_selectedDifficulty.value == difficulty) null else difficulty
        applyFilters()
        updateHasActiveFilters()
    }

    fun clearFilters() {
        _selectedSpirits.value = emptySet()
        _selectedCategories.value = emptySet()
        _selectedDifficulty.value = null
        _selectedMood.value = null
        applyFilters()
        updateHasActiveFilters()
    }

    /**
     * Mood-based discovery. Selecting a mood filters the catalog by tags and
     * reports the preference to Purchasely for audience targeting.
     */
    fun selectMood(mood: CocktailMood?) {
        val newMood = if (_selectedMood.value == mood) null else mood
        _selectedMood.value = newMood
        if (newMood != null) {
            // PURCHASELY: track the user's drinking mood as a custom attribute. Console-side
            // audiences can target e.g. preferred_mood == "zero_proof" with a dedicated screen.
            // Docs: https://docs.purchasely.com/advanced-features/user-attributes
            purchaselyWrapper.setUserAttribute("preferred_mood", newMood.key)
        }
        applyFilters()
        updateHasActiveFilters()
    }

    /**
     * "Surprise me" — picks a random cocktail from the currently filtered list.
     * Returns its id for navigation, or null when the list is empty.
     */
    fun onSurpriseMe(): String? {
        val candidates = _cocktails.value.ifEmpty { return null }
        // PURCHASELY: count how often the user asks for a surprise. A console campaign can
        // trigger a dedicated screen after N uses (engaged-user segmentation).
        purchaselyWrapper.incrementUserAttribute("surprise_me_count")
        return candidates.random().id
    }

    fun onPresentationDismissed() {
        premiumRepository.refreshPremiumStatus()
    }

    private fun applyFilters() {
        _cocktails.value = getFilteredCocktails(
            query = _searchQuery.value,
            spirits = _selectedSpirits.value,
            categories = _selectedCategories.value,
            difficulty = _selectedDifficulty.value,
            mood = _selectedMood.value,
        )
    }

    companion object {
        private const val TAG = "HomeViewModel"
    }
}

File: ios/Shaker/Screens/Home/HomeViewModel.swift
import UIKit

@MainActor
class HomeViewModel: ObservableObject {

    @Published var cocktails: [Cocktail] = []
    @Published var searchQuery = "" {
        didSet {
            if !searchQuery.isEmpty {
                wrapper.setUserAttribute(true, forKey: "has_used_search")
            }
            applyFilters()
        }
    }

    // Filter state
    @Published var selectedSpirits: Set<String> = [] {
        didSet { applyFilters() }
    }
    @Published var selectedCategories: Set<String> = [] {
        didSet { applyFilters() }
    }
    @Published var selectedDifficulty: String? {
        didSet { applyFilters() }
    }
    @Published var selectedMood: CocktailMood? {
        didSet { applyFilters() }
    }

    // Prefetched presentations
    @Published var inlinePresentation: FetchResult?
    @Published var filtersPresentation: FetchResult?
    @Published var isFiltersLoading = false

    private let repository: CocktailRepository
    private let wrapper: PurchaselyWrapping
    private let allCocktails: [Cocktail]

    var availableSpirits: [String] { repository.spirits() }
    var availableCategories: [String] { repository.categories() }
    var availableDifficulties: [String] { repository.difficulties() }

    var hasActiveFilters: Bool {
        !selectedSpirits.isEmpty || !selectedCategories.isEmpty
            || selectedDifficulty != nil || selectedMood != nil
    }

    init(repository: CocktailRepository = .shared,
         wrapper: PurchaselyWrapping = PurchaselyWrapper.shared) {
        self.repository = repository
        self.wrapper = wrapper
        self.allCocktails = repository.allCocktails()
        cocktails = allCocktails
    }

    private func applyFilters() {
        cocktails = allCocktails.filter { cocktail in
            let matchesQuery = searchQuery.isEmpty || cocktail.name.localizedCaseInsensitiveContains(searchQuery)
            let matchesSpirit = selectedSpirits.isEmpty || selectedSpirits.contains(cocktail.spirit)
            let matchesCategory = selectedCategories.isEmpty || selectedCategories.contains(cocktail.category)
            let matchesDifficulty = selectedDifficulty == nil || cocktail.difficulty == selectedDifficulty
            let matchesMood = selectedMood == nil || selectedMood!.matches(cocktail)
            return matchesQuery && matchesSpirit && matchesCategory && matchesDifficulty && matchesMood
        }
    }

    // MARK: - Mood-based discovery

    /// Selecting a mood filters the catalog by tags and reports the preference
    /// to Purchasely for audience targeting. Mirrors the Android `selectMood`.
    func selectMood(_ mood: CocktailMood?) {
        let newMood = selectedMood == mood ? nil : mood
        selectedMood = newMood
        if let newMood {
            // PURCHASELY: track the user's drinking mood as a custom attribute. Console-side
            // audiences can target e.g. preferred_mood == "zero_proof" with a dedicated screen.
            // Docs: https://docs.purchasely.com/advanced-features/user-attributes
            wrapper.setUserAttribute(newMood.key, forKey: "preferred_mood")
        }
    }

    // MARK: - Surprise me

    /// "Surprise me" — picks a random cocktail from the currently filtered list.
    /// Returns its id for navigation, or nil when the list is empty.
    /// Mirrors the Android `onSurpriseMe`.
    func surpriseMe() -> String? {
        guard let pick = cocktails.randomElement() else { return nil }
        // PURCHASELY: count how often the user asks for a surprise. A console campaign can
        // trigger a dedicated screen after N uses (engaged-user segmentation).
        wrapper.incrementUserAttribute(forKey: "surprise_me_count")
        return pick.id
    }

    func prefetchPresentations(isPremium: Bool) {
        guard !isPremium else { return }

        Task { @MainActor in
            isFiltersLoading = true
            filtersPresentation = await wrapper.loadPresentation(placementId: "filters") { result in
                if case .purchased = result { PremiumManager.shared.refreshPremiumStatus() }
                if case .restored = result { PremiumManager.shared.refreshPremiumStatus() }
            }
            isFiltersLoading = false
        }

        Task { @MainActor in
            inlinePresentation = await wrapper.loadPresentation(placementId: "inline") { result in
                if case .purchased = result { PremiumManager.shared.refreshPremiumStatus() }
                if case .restored = result { PremiumManager.shared.refreshPremiumStatus() }
            }
        }
    }

    func displayFiltersPaywall(from viewController: UIViewController?) {
        switch filtersPresentation {
        case .success(let handle):
            wrapper.display(handle: handle, from: viewController)
        case .error(let error):
            // PURCHASELY: the prefetch failed (e.g. offline at launch). Log and retry it so
            // the paywall becomes available instead of leaving the button dead forever.
            print("[Shaker] Filters paywall unavailable (\(error?.localizedDescription ?? "unknown")), retrying prefetch")
            retryFiltersPrefetch()
        case .deactivated, .client:
            print("[Shaker] Filters paywall not displayable: \(String(describing: filtersPresentation))")
        case nil:
            print("[Shaker] Filters paywall still loading")
        }
    }

    private func retryFiltersPrefetch() {
        guard !isFiltersLoading else { return }
        Task { @MainActor in
            isFiltersLoading = true
            filtersPresentation = await wrapper.loadPresentation(placementId: "filters") { result in
                if case .purchased = result { PremiumManager.shared.refreshPremiumStatus() }
                if case .restored = result { PremiumManager.shared.refreshPremiumStatus() }
            }
            isFiltersLoading = false
        }
    }

    func toggleSpirit(_ spirit: String) {
        if selectedSpirits.contains(spirit) {
            selectedSpirits.remove(spirit)
        } else {
            selectedSpirits.insert(spirit)
        }
    }

    func toggleCategory(_ category: String) {
        if selectedCategories.contains(category) {
            selectedCategories.remove(category)
        } else {
            selectedCategories.insert(category)
        }
    }

    func selectDifficulty(_ difficulty: String) {
        selectedDifficulty = selectedDifficulty == difficulty ? nil : difficulty
    }

    func clearFilters() {
        selectedSpirits = []
        selectedCategories = []
        selectedDifficulty = nil
        selectedMood = nil
    }
}
